src/mlj_model_interface.jl

```diff
@@ -53,74 +55,61 @@ end
 
 
 # # FIT AND  UPDATE
+include("fit_utils.jl")
+include("entity_embedding_utils.jl")
 
-const ERR_BUILDER =
-    "Builder does not appear to build an architecture compatible with supplied data. "
+const ERR_BUILDER = "Builder does not appear to build an architecture compatible with supplied data. "
 
 true_rng(model) = model.rng isa Integer ? Random.Xoshiro(model.rng) : model.rng
 
-# Models implement L1/L2 regularization by chaining the chosen optimiser with weight/sign
-# decay.  Note that the weight/sign decay must be scaled down by the number of batches to
-# ensure penalization over an epoch does not scale with the choice of batch size; see
-# https://github.com/FluxML/MLJFlux.jl/issues/213.
-
-function regularized_optimiser(model, nbatches)
-    model.lambda == 0 && return model.optimiser
-    λ_L1 = model.alpha*model.lambda
-    λ_L2 = (1 - model.alpha)*model.lambda
-    λ_sign = λ_L1/nbatches
-    λ_weight = 2*λ_L2/nbatches
-
-    # recall components in an optimiser chain are executed from left to right:
-    if model.alpha == 0
-        return Optimisers.OptimiserChain(
-            Optimisers.WeightDecay(λ_weight),
-            model.optimiser,
-        )
-    elseif model.alpha == 1
-        return Optimisers.OptimiserChain(
-            Optimisers.SignDecay(λ_sign),
-            model.optimiser,
-        )
-   else  return Optimisers.OptimiserChain(
-        Optimisers.SignDecay(λ_sign),
-        Optimisers.WeightDecay(λ_weight),
-        model.optimiser,
-        )
-    end
-end
 
 function MLJModelInterface.fit(model::MLJFluxModel,
-                               verbosity,
-                               X,
-                               y)
-
+    verbosity,
+    X,
+    y)
+    # GPU and rng related variables
     move = Mover(model.acceleration)
-
     rng = true_rng(model)
-    shape = MLJFlux.shape(model, X, y)
 
-    chain = try
-        build(model, rng, shape) |> move
-    catch ex
-        @error ERR_BUILDER
-        rethrow()
+    # Get input properties
+    shape = MLJFlux.shape(model, X, y)
+    cat_inds = get_cat_inds(X)
+    pure_continuous_input = (length(cat_inds) == 0)
+
+    # Decide whether to enable entity embeddings (e.g., ImageClassifier won't)
+    enable_entity_embs = is_embedding_enabled_type(typeof(model)) && !pure_continuous_input
+
+    # Prepare entity embeddings inputs and encode X if entity embeddings enabled
+    if enable_entity_embs
+        X = convert_to_table(X)
+        featnames = Tables.schema(X).names
+        entityprops, entityemb_output_dim =
+            prepare_entityembs(X, featnames, cat_inds, model.embedding_dims)
+        X, ordinal_mappings = ordinal_encoder_fit_transform(X; featinds = cat_inds)
     end
```

Comment:
It's probably not critical here, so you can leave it, but generally, in a function you  should use a new variable name if you *change the type* of the object, as in lines 84 (and 88?). At least that  [appears](https://docs.julialang.org/en/v1/manual/performance-tips/#Avoid-changing-the-type-of-a-variable) to be Julia's advice.